Moved pod_update to a pod ginkgo test suite. #3941

rrati opened the pull request.

satnam6502 commented:
@filbranden

satnam6502 commented:
Kicking Travis to restart.

filbranden commented:
Can we keep the same file name and the same test name "TestPodUpdate" ?

The same filename would make it easier to review the code to see what was kept and what was changed...

The name on Describe() is what we're indexing our tests with in Jenkins and also for the `cmd/e2e -t` parameter, so I'd really prefer if we didn't change test names while converting them...

Thanks!
Filipe

rrati commented:
I changed the file name because a ginkgo test suite usually isn't just 1 
test and I was creating a file for all the specs that will test pods. 
That's also why I used a different name in the Describe() line.  Is the 
intention not to move to that format?

There's actually 2 tests in the new file instead of just the singular 
one that was in the original file.  I wouldn't say the added test 
(submit and remove) is really adding anything that the update test 
doesn't cover though.  It probably makes sense to just remove that test.

Is it possible to use something else other than the Describe() entry for 
test indexing?  Maybe that's a later step in the conversion?

Rob

On 01/29/2015 03:01 PM, Filipe Brandenburger wrote:
> Can we keep the same file name and the same test name "TestPodUpdate" ?
>
> The same filename would make it easier to review the code to see what
> was kept and what was changed...
>
> The name on Describe() is what we're indexing our tests with in Jenkins
> and also for the |cmd/e2e -t| parameter, so I'd really prefer if we
> didn't change test names while converting them...
>
> Thanks!
> Filipe
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/3942#issuecomment-72093567>.
>

filbranden commented:
@rrati Yeah let's do one step at a time please!

We're barely having the ability to collect these JUnit XML reports in Jenkins now, before we talk about renaming tests I want to make sure we're getting useful data from Jenkins. So please keep the test name in Describe() the same as it is now.

Also, if you want to rename files, do that in a separate step. If they files look enough alike git will detect it's a rename, but if you rename it *and* change a bunch of things in the same commit then it will have a harder time figuring it out...

Baby steps... baby steps.

zmerlynn commented:
Actually, Jenkins has no trouble collecting multiple suites right now, unless you think there's some issue in driver.go with outputting them? I really don't mind the conversion being to the final resting place.

zmerlynn commented:
FWIW, Jenkins glommed far more onto the string "Kubernetes e2e Suite", so the "description" string out of `RunSpecs` seems to dictate the intermediate level of its reporting. The "suite" aspect in Gingko's JUnit reporting somehow (we might want to look into this). Regardless, this structure looks fine, and this test actually splits things up into two assertions, which is cool.

LGTM. Kicking Travis.

filbranden commented:
Please don't merge it as is.

At the very least, the two commits should be squashed together.

I'm also strongly against renaming the file on this same commit. I really want GitHub to detect a file rename and the best way to do it is in its own commit (you could do it before the conversion or after the conversion, that's about the same.)

I'm against the TestPodUpdate -> Pod rename but if you feel strongly about doing the rename then that's fine. But then instead of doing "one off" renames and having old-style and new-style living together for a long time I'd rather have a single rename of everything on its own commit. You could do it now before the conversion if you'd like.

Cheers,
Filipe

rrati commented:
I've no problems creating a separate PR for changing the name of the file if that's required.  I think a discussion of the end goal and the steps we'd like to take to get there might be in order.  My assumption has been this ginkgo transition would result in a multiple specs in a single file, ie test suites.  In that situation having "TestPodUpdate" as the name of the suite is misleading since the suite will be testing more than just pod update.  I think a quick hangout could help clarify the steps and end goals here.

zmerlynn commented:
At the very least you might want an initial, separate commit that does the rename. It's possible that if you do the git mv in one commit and the changes in another that they can still sit in the same PR.

I actually have nothing against this PR. I'd like to cleanup the file tracking so history isn't lost (so agree with Filipe there), and it looks like Travis is still red (squash your gofmt piece back in?).

Maybe IRC later?

roberthbailey commented on test/e2e/pod.go:
```diff
@@ -0,0 +1,122 @@
+var _ = Describe("Pods", func() {
+	var (
+		c   *client.Client
+		err error
```
Why is err a test-wide var?

roberthbailey commented on test/e2e/pod.go:
```diff
@@ -0,0 +1,122 @@
+		if err != nil {
+			Fail(fmt.Sprintf("Failed to create pod: %v", err))
+		}
+		defer podClient.Delete(pod.Name)
```
The should update a pod test below also calls defer GinkgoRecover() inside of this defer. The two specs should be consistent.

roberthbailey commented on test/e2e/pod.go:
```diff
@@ -0,0 +1,122 @@
+		c = loadClientOrDie()
+	})
+
+	It("should submit and remove a pod", func() {
```
The first three steps (load, submit to k8s, verify) are exactly the same code between these two tests. It seems like they should be extracted into a common helper so prevent drift (which we already have with the different defer functions to delete the pod).

Commit:
Renamed the pod_update file to pods #3941

Commit:
Converted TestPodUpdate to native ginkgo syntax #3941

Commit:
Added a submit/remove test to pods #3941

Commit:
Converted TestPodHasServiceEnvVars to native ginkgo syntax #3941

Commit:
Moved TestPodHasServiceEnvVars into the pods test suite and deleted the
original file #3941

filbranden commented:
I like what I see here.

Looks like merging #3995 caused a conflict on this one... sorry about that. Would you like to rebase it and bring those changes in? If you'd like, I can do a manual merge of this branch and resolve the conflicts as I do that.

@roberthbailey Do you still have comments on the code? Are you satisfied with the resolution of the ones you had? I think there's still quite a bit of code duplication here but I think I'd be OK merging this one as is now and trying to polish the code in future PRs to prevent us from getting on a standstill with this one...

Cheers,
Filipe

roberthbailey commented:
Sorry, I didn't see that my comments had been addressed. Looking now.

roberthbailey commented on test/e2e/pods.go:
```diff
@@ -0,0 +1,259 @@
+			  "id": "srv",
+			  "containers": [{
+				"name": "srv",
+				"image": "kubernetes/serve_hostname",
```
Does this image have a specific tag we can reference?

satnam6502 replied:
Yes it does. I think the current one is :1.1 and I encourage its use.
Note that serve_hostname (not tag, or :latest) is the oldest image and adds a newline character to the newline.
Note that serve_hostname:1.0 (based on ubuntu so big) or server_hostname:1.1 (based on scratch, so small) **do not** add a newline character after the hostname.

roberthbailey commented on test/e2e/pods.go:
```diff
@@ -0,0 +1,259 @@
+				"image": "kubernetes/serve_hostname",
+				"ports": [{
+				  "containerPort": 9376,
+				  "hostPort": 8080
```
Do we really need to use hostPort here?

roberthbailey commented:
I agree that there is still a lot of code duplication that can be cleaned up, but I'll defer to you on whether that should block this getting in. I'm happy with the direction this is taking us (multiple specs per test and making the test groupings make logical sense).

rrati commented:
@roberthbailey I addressed some of the issues you raised earlier but didn't want to differ from the originals too much code in this CR given the nature of the changes (filenames, moving content, etc).  I agree with all the comments you made.

rrati commented:
@filbranden I'll rebase and push again

rrati commented:
@filbranden Rebased branch pushed

filbranden commented:
Thanks @rrati! Please keep the contributions coming!

roberthbailey commented:
No worries. Looking forward to seeing them addressed in a future PR.

satnam6502 commented:
+1. Well, +N hopefully. Thanks.

rrati commented:
https://github.com/GoogleCloudPlatform/kubernetes/issues/3837
